src/RenderManager.cpp

```diff
@@ -24,9 +24,14 @@ void RenderManager::FreeImageErrorHandler(FREE_IMAGE_FORMAT fif, const char *mes
 	printf(" ***\n\n");
 }
 
-RenderManager::RenderManager(int mapTileWidth, int mapTileHeight, int bpp, int scaleFactor) : 
+RenderManager::RenderManager(unsigned mapTileWidth, unsigned mapTileHeight, int bpp, unsigned scaleFactor) : 
 	scaleFactor(scaleFactor),
-	freeImageBmpDest(mapTileWidth * scaleFactor, mapTileHeight * scaleFactor, bpp) { }
+	freeImageBmpDest(mapTileWidth * scaleFactor, mapTileHeight * scaleFactor, bpp) 
+{
+	if (mapTileWidth * mapTileHeight > std::numeric_limits<unsigned int>::max() / scaleFactor - scaleFactor) {
```

Comment:
The product `mapTileWidth * mapTileHeight` could overflow. More to the point, I think the check involving `scaleFactor` was meant to work on each dimension individually. As in, the check is against overflow in each of the two multiplies done in:
```cpp
freeImageBmpDest(mapTileWidth * scaleFactor, mapTileHeight * scaleFactor, bpp)
```

Actually, scratch that, it's not about the destination image dimensions. It's about the scaled source tileset dimensions. The check being removed was on `tileIndex`.

Reply:
I was trying to check the final tileset index by translating it into a pixel coordinate and ensuring that it doesn't overflow. 

Maybe this doesn't fit the bill though. I'm having trouble understanding the checks to explain to you.

Thanks for looking at it so quickly.

-Brett

Reply:
Right. The tilesets have a fixed tile size of 32x32 pixels. They are a single tile wide, and a variable number of tiles high. That means the width is never really an issue. The original source has a width of `32` pixels, while the scaled copy has a width of `scaleFactor`. Assuming `scaleFactor` is the same data type as the width, there can be no overflow there.

The check is on the `tileIndex`, which must be in range for what index values are allowed, and also must be restricted such that `tileIndex * scaleFactor + scaleFactor - 1` does not overflow. That is, the max Y value at the bottom scanline of the last tile (where Y grows going down), must be in a representable range. And maybe we want to remove the ` - 1` part for ease of bounds checking.

When creating the scaled copy of the tileset, we can bounds check the max Y value to ensure there is no overflow. That is, `tileCount * scaleFactor` scanlines can be represented, with coordinates 0..(tileCount * scaleFactor - 1).

When pasting the tile, we can do a bounds check on `tileIndex`, to make sure it's within the bounds of the tileset. Assuming that is true, then the pre-checked scaled tileset size will ensure that `tileIndex * scaleFactor + scaleFactor` is also in bounds, because `tileCount * scaleFactor` was in bounds, and `tileIndex < tileCount` (0 <= tileIndex < tileCount).

----

From another perspective, assuming image dimensions fall in a representable range, the maximum supported tiles in a tileset is `tileCount <= maxY / scaleFactor` (using default round down in case `maxY` is not an integer multiple of `scaleFactor`).

Reply:
Implemented. I think this is being checked in the proper place. The downside is that the freeImageBmp is created before the check is performed. So there is a good chance that freeImage will get mad and return an error before we catch.

Reply:
I'm looking at this again, and realizing the check being removed doesn't match up with the check being added. The original check being removed was on reading the source scaled tileset images. This new check is on the dimensions of the destination map image.

Further, the check being removed actually needs to be simplified. It should be checking for a valid tile index. The pixel dimensions can be pre-checked when the tilesets are scaled, so that only the index needs to be checked while pasting.